packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+
+  /// The height from the [Scaffold]'s top where its body starts.
+  final double contentTop;
+
+  /// The minimum horizontal padding the [FloatingActionButton] should observe.
+  final double horizontalFabPadding;
+
+  /// The [Size] of the whole [Scaffold].
+  final Size scaffoldSize;
+
+  /// The [Size] of the [Scaffold]'s [SnackBar] (if available).
+  /// 
+  /// The Scaffold will determine the snackBarSize if the Scaffold
+  /// is currently showing a [SnackBar].
+  final Size snackBarSize;
+
+  /// The [Scaffold]'s [TextDirection].
+  final TextDirection textDirection;
+}
+
+class _CenterFloatFab extends FabPositioner {
+  const _CenterFloatFab();
+
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    // Compute the x-axis offset.
+    final double fabX = (scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.fabSize.width) / 2;
+
+    // Compute the y-axis offset.
+    final double contentBottom = scaffoldGeometry.contentBottom;
+    final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
+    final double fabHeight = scaffoldGeometry.fabSize.height;
+    final double snackBarHeight = scaffoldGeometry.snackBarSize.height;
+    double fabY = contentBottom - fabHeight - _kFloatingActionButtonMargin;
+    if (snackBarHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - snackBarHeight - fabHeight - _kFloatingActionButtonMargin);
+    if (bottomSheetHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);
+
+    return new Offset(fabX, fabY);
+  }
+}
+
+class _EndFloatFab extends FabPositioner {
+
+  const _EndFloatFab();
+
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    // Compute the x-axis offset.
+    double fabX;
+    assert(scaffoldGeometry.textDirection != null);
+    switch (scaffoldGeometry.textDirection) {
+      case TextDirection.rtl:
+        fabX = _kFloatingActionButtonMargin + scaffoldGeometry.horizontalFabPadding;
+        break;
+      case TextDirection.ltr:
+        fabX = scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.fabSize.width - _kFloatingActionButtonMargin - scaffoldGeometry.horizontalFabPadding;
+      break;
+    }
+
+    // Compute the y-axis offset.
+    final double contentBottom = scaffoldGeometry.contentBottom;
+    final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
+    final double fabHeight = scaffoldGeometry.fabSize.height;
+    final double snackBarHeight = scaffoldGeometry.snackBarSize.height;
+
+    double fabY = contentBottom - fabHeight - _kFloatingActionButtonMargin;
+    if (snackBarHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - snackBarHeight - fabHeight - _kFloatingActionButtonMargin);
+    if (bottomSheetHeight > 0.0)
+      fabY = math.min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);
+
+    return new Offset(fabX, fabY);
+  }
+}
+// Examples can assume:
+// ScaffoldGeometry scaffoldGeometry;
```

Comment:
we normally put these pragmas at the top of the file after the imports.

Reply:
I've pulled this comment out.